### ISSUE TITLE
why push the vasa loss into the stage 2?

### ISSUE THREAD

**Issue by Jason-Chi-xx** [johndpope/VASA-1-hack#33]
First of all, appreciate for your good work. I want to ask the reason that pushing the vasa loss into the training stage 2. Is it because there are already many losses in stage 1, adding vasa loss to it may make the training difficult to converge? Looking forward to your apply, thanks

**Comment by johndpope** [comment]
i did a lot of work on stage 2 - (rebuilt maybe 5 or 6 times) not released - 

the author of emoportaits worked on megaportraits - here's my fork
https://github.com/johndpope/nemo


the latent space of the holistic transformer is so vast that the convergence is taking a long time....
i bought the 5090 - i think maybe 16 weeks delay....
https://drive.google.com/drive/u/0/folders/1wiGA0g-QA0tKL-eHnNt5UJ97LVhekIxg


essentially - we're matching the audio + expression embedding

they did this here
https://github.com/DiffPoseTalk/DiffPoseTalk/blob/main/models/diff_talking_head.py


and more recently with joyvasa - liveportrait
https://github.com/jdh-algo/JoyVASA/blob/main/src/modules/dit_talking_head.py

i start pulling apart joyvasa and it's training - and again it's really slow to converge on my 3090

this expression embedding drives me up the wall  
https://wandb.ai/snoozie/vasa/runs/2h3iwmmn?nw=nwusersnoozie


![Image](https://github.com/user-attachments/assets/c116c65c-2b9e-411c-9be2-19e10dd19535)

https://github.com/user-attachments/assets/0d797ee6-6eec-4b43-a38e-47d1ed057ee9


i create a new algorithm to potentially speed up convergence for expression and possibly even the audio 
https://github.com/johndpope/HiDETR/

you can run https://github.com/johndpope/HiDETR/blob/main/python/main.py
it will spit out some pretty graphs. the difference gets smaller and smaller in recreation.
what this means is -  if you have 40,000 videos - with many different facial expressions - the 128 dim from emo - needs be recreated perfectly - if it's off - you'll get distorted heads like my google drive videos. this hidetr - will cut up the latent space into location sensitive hashes. but now im wondering whether they should be categorised into facial action units FACS to guide the model more...

this video has overview of deformable detr - (this stuff is working in 2d) 
https://www.youtube.com/watch?v=9UG4amweIjk&t=87s&ab_channel=MakGaiduk



I'd suggest training this joyvasa - it's slow - the curse of dimensionality 
- but if what my HiDETR could be the blessing of dimensionality.... claude is saying its 15x faster - that's on the 128dim - could be 3/4x on 512 dim.  
https://github.com/johndpope/JoyVASA/commit/e082c436e2fe775350a08c9365c886db4c4036b5

**Comment by Jason-Chi-xx** [comment]
Thanks for your reply, it is inspiring to see these works you have done. But I am also very curious about the reason that moving the vasa loss from stage1 to stage2. By doing so, can I just deem the stage 1 as a megaportrait model?

**Comment by johndpope** [comment]
sorry - this is the vasa loss - 
and it does belong in stage 1 - it shouldn't be removed here.
https://github.com/johndpope/MegaPortrait-hack/blob/main/train.py#L259 

its been a while since i looked at this - 
frankly when i was playing around with megaportraits - i couldnt get anywhere without  Kevin's Fringe's consistency code - fake / real. there's something wrong with it - at the time i skip over it - people in megaportrait repo said that they had id leakage - and it was a bit elusive - i don't know if its a) the losses b) some batch normalization (images from batch somehow screwing things up) c) lack of training / data

also my code /i overlooked  modnet - which in megaportraits they say is big factor - 
https://github.com/johndpope/MegaPortrait-hack/issues/59


if this cross identity is fixable -(it is in emo portraits already)
1) i would add the modnet - 
2) carefully extract the loss from emo (there's so many) 
https://github.com/neeek2303/EMOPortraits/tree/main/losses
3) use emo's volumetric model as a pseudo ground truth - its already trained and it matches the dims here
would help eliminate variables https://github.com/johndpope/VASA-1-hack/issues/28



IdtEmbed
https://github.com/neeek2303/EMOPortraits/blob/93cd6a565ca611e13fe75fd98ef198a737fba9c4/networks/volumetric_avatar/identity_embedder.py#L28


https://github.com/johndpope/MegaPortrait-hack/blob/580cab3cadc1873dad5e7b9af0a0d9e882dee486/train.py#L142

when i look at the emo code - there's so many args upon args upons
(in my fork i collapse into yaml file with 300 args)
https://github.com/johndpope/nemo/blob/main/infer.py#L87


i did ask for help here - 
https://github.com/neeek2303/EMOPortraits/issues/27

**Comment by Jason-Chi-xx** [comment]
Thanks for your reply, hope everything goes well with your work

**Comment by johndpope** [comment]
i push code -
